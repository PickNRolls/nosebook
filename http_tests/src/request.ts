import supertest, { Test } from 'supertest';
import TestAgent from 'supertest/lib/agent';

const HOST = 'http://backend:8080';
const SESSION_HEADER_KEY = 'X-Auth-Session-Id';

class TestAgentWrapper {
  private host: string;
  private agent: TestAgent;

  public constructor(host: string) {
    this.host = host;
    this.agent = supertest(host);
  }

  public unwrap(): TestAgent {
    return this.agent;
  }

  public extend(prefixUrl: string): TestAgentWrapper {
    return new TestAgentWrapper(this.host + prefixUrl);
  }

  private hook<M extends 'get' | 'post'>(method: M, url: string): Test {
    const test = this.agent[method](url);

    test.set(SESSION_HEADER_KEY, 'bb23af03-be50-4bce-b729-b259b2e02e54');

    return test;
  }

  public get(url: string): Test {
    return this.hook('get', url);
  }

  public post(url: string): Test {
    return this.hook('post', url);
  }
};

export default new TestAgentWrapper(HOST);


