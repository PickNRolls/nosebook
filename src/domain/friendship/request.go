package domainfriendship

import (
	"nosebook/src/errors"
	"time"

	"github.com/google/uuid"
)

type FriendRequest struct {
	RequesterId uuid.UUID `json:"requesterId" db:"requester_id"`
	ResponderId uuid.UUID `json:"responderId" db:"responder_id"`
	Message     string    `json:"message" db:"message"`
	Accepted    bool      `json:"accepted" db:"accepted"`
	Viewed      bool      `json:"viewed" db:"viewed"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`

	events []Event
}

func New(
	requesterId uuid.UUID,
	responderId uuid.UUID,
	message string,
	accepted bool,
	viewed bool,
	createdAt time.Time,
	raiseCreatedEvent bool,
) *FriendRequest {
	output := &FriendRequest{
		RequesterId: requesterId,
		ResponderId: responderId,
		Message:     message,
		Accepted:    accepted,
		Viewed:      viewed,
		CreatedAt:   createdAt,
		events:      make([]Event, 0),
	}

	if raiseCreatedEvent {
		output.raiseEvent(NewCreatedEvent())
	}

	return output
}

func (this *FriendRequest) raiseEvent(event Event) {
	this.events = append(this.events, event)
}

func (this *FriendRequest) Events() []Event {
	return this.events
}

func (this *FriendRequest) ViewBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отметить заявку просмотренной")
	}

	this.Viewed = true
	this.raiseEvent(NewViewedEvent())

	return nil
}

func (this *FriendRequest) AcceptBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может принять заявку")
	}

	this.Accepted = true
	this.Viewed = true
	this.raiseEvent(NewAcceptedEvent())

	return nil
}

func (this *FriendRequest) DenyBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отклонить заявку")
	}

	this.Accepted = false
	this.Viewed = true
	this.raiseEvent(NewDeniedEvent())

	return nil
}

func (this *FriendRequest) RemoveBy(userId uuid.UUID) *errors.Error {
	if !this.Accepted {
		return errors.New("Friendship Error", "Нельзя удалить непринятую заявку")
	}

	this.Accepted = false
	previousRequesterId := this.RequesterId
	previousResponderId := this.ResponderId

	if this.RequesterId == userId {
		this.RequesterId, this.ResponderId = this.ResponderId, this.RequesterId
	}

	this.raiseEvent(NewRemovedEvent(previousRequesterId, previousResponderId))

	return nil
}
