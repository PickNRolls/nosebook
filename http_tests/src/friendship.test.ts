import request from './request';

describe('friendship', () => {
  let friends = request.extend({
    prefixUrl: '/friendship'
  });

  test('POST /send-request', async () => {
    let response = await friends
      .post('/send-request')
      .send({
        responderId: '48683858-796c-45ad-a361-9e3d6d003354',
        message: 'test add',
      })
      .expect(200);

    expect(response.body).toMatchSnapshot();
  });

  test('POST /accept-request', async () => {
    let response = await friends
      .post('/accept-request')
      .send({
        requesterId: '1ae02f69-ea1a-4308-b825-0e5896e652e4',
      })
      .expect(200);

    expect(response.body).toMatchSnapshot();
  });

  test('POST /deny-request', async () => {
    let response = await friends
      .post('/deny-request')
      .send({
        requesterId: 'baa0e8bc-385f-4314-9580-29855aff2229',
      })
      .expect(200);

    expect(response.body).toMatchSnapshot();
  });

  test('POST /remove-friend', async () => {
    let response = await friends
      .post('/remove-friend')
      .send({
        friendId: '2db640fd-7aa4-4bba-8ee6-3935b700297a',
      })
      .expect(200);

    expect(response.body).toMatchSnapshot();
  });

  describe('GET /', () => {
    test('filter friends for userId', async () => {
      let response = await friends
        .get('/')
        .query({
          userId: 'ed1a3fd0-4d0b-4961-b4cd-cf212357740d',
          accepted: true,
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    test('filter friends with limit', async () => {
      let response = await friends
        .get('/')
        .query({
          userId: 'ed1a3fd0-4d0b-4961-b4cd-cf212357740d',
          accepted: true,
          limit: 1,
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();

      response = await friends
        .get('/')
        .query({
          userId: 'ed1a3fd0-4d0b-4961-b4cd-cf212357740d',
          limit: 1,
          accepted: true,
          next: response.body.data.next,
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    test('filter online friends', async () => {
      let response = await friends
        .get('/')
        .query({
          userId: 'ed1a3fd0-4d0b-4961-b4cd-cf212357740d',
          onlyOnline: true,
          accepted: true,
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    test('filter incoming pending requests', async () => {
      let response = await friends
        .get('/')
        .query({
          userId: 'ed1a3fd0-4d0b-4961-b4cd-cf212357740d',
          onlyIncoming: true,
          accepted: false,
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    test('filter outcoming pending requests', async () => {
      let response = await friends
        .get('/')
        .query({
          userId: 'ed1a3fd0-4d0b-4961-b4cd-cf212357740d',
          onlyOutcoming: true,
          accepted: false,
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });
});

