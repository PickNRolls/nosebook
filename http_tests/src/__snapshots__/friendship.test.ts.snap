// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`friendship POST /accept-request 1`] = `
{
  "data": {
    "accepted": true,
    "createdAt": "2024-02-16T15:36:38Z",
    "message": "test request",
    "requesterId": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
    "responderId": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
    "viewed": true,
  },
  "ok": true,
}
`;

exports[`friendship POST /deny-request 1`] = `
{
  "data": {
    "accepted": false,
    "createdAt": "2024-02-16T15:36:38Z",
    "message": "test request",
    "requesterId": "baa0e8bc-385f-4314-9580-29855aff2229",
    "responderId": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
    "viewed": true,
  },
  "ok": true,
}
`;

exports[`friendship POST /remove-friend 1`] = `
{
  "data": {
    "accepted": false,
    "createdAt": "2024-02-16T15:36:38Z",
    "message": "test",
    "requesterId": "2db640fd-7aa4-4bba-8ee6-3935b700297a",
    "responderId": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
    "viewed": true,
  },
  "ok": true,
}
`;

exports[`friendship POST /send-request 1`] = `
{
  "data": {
    "accepted": false,
    "createdAt": "2024-08-10T10:10:10.00000001Z",
    "message": "test add",
    "requesterId": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
    "responderId": "48683858-796c-45ad-a361-9e3d6d003354",
    "viewed": false,
  },
  "ok": true,
}
`;
